colony_picker/inverse_kinematics.py

```diff
@@ -0,0 +1,634 @@
+            describe every joint on the robot arm in the format
+            [[theta offset 1, theta offset 2, ... , theta offset n],
+            [alpha 1, alpha 2, ... , alpha n],
+            [a 1, a 2, ... , a n], 
+            [d 1, d 2, ... , d n]]
+            where n is the number of joints on the robot arm. 
+
+    Returns:
+        np.ndarray: an array containing 4 x 4 transformation matrices that each
+            describe the position and orientation of each joint relative to the
+            base frame in the format [t mat 1, t mat 2, ... , t mat num joints].
+    """
+    num_joints = len(dh_params[0])
+    t_mats = []
+    accumulator_t_mat = np.identity(4)
+    for joint_idx in range(num_joints):
+        t_mat = get_t_mat(thetas[joint_idx], dh_params[:, joint_idx])
+        accumulator_t_mat = np.matmul(accumulator_t_mat, t_mat)
+        t_mats.append(accumulator_t_mat)
+
+    return t_mats
+
+
+def get_end_effector_pose(thetas: np.ndarray, dh_params: np.ndarray,
+                          euler: bool = False):
+    """Conducts forward kinematics on a robot arm described by dh_params,
+    outputting the end effector pose given a set of joint angles described by
+    thetas. 
+
+    Args:
+        thetas (np.ndarray): a 1 x n array of joint angles in the format
+            [joint angle 1, joint angle 2, ... , joint angle n] in which n is 
+            the number of joints on the robot arm. 
+        dh_params (np.ndarray): a 4 x n array containing the dh parameters to 
+            describe every joint on the robot arm in the format
+            [[theta offset 1, theta offset 2, ... , theta offset n],
+            [alpha 1, alpha 2, ... , alpha n],
+            [a 1, a 2, ... , a n], 
+            [d 1, d 2, ... , d n]]
+            where n is the number of joints on the robot arm. 
+        euler (bool, optional): whether the end effector pose should
+            be returned with its orientation described in ZYX euler angles. 
+            Defaults to False which implies orientation is described in
+            quaternions.
+
+    Returns:
+        np.ndarray: an array describing the position and orientation of the end
+            effector given thetas and dh_params. If euler is true, the array is
+            size 1 x 6, if not, the orientation is described by quaternions, 
+            making the array size 1 x 7. 
+    """
+    t_mats = get_t_mats(thetas, dh_params)
+    end_effector_position = t_mats[-1][:3, 3]
+    rot_mat = t_mats[-1][:3, :3]
+    if euler:
+        end_effector_rotation = get_euler_from_rot_mat(rot_mat)
+        end_effector_pose = np.zeros(6)
+    else:
+        end_effector_rotation = get_quat_from_rot_mat(rot_mat)
+        end_effector_pose = np.zeros(7)
+    end_effector_pose[:3] = end_effector_position
+    end_effector_pose[3:] = end_effector_rotation
+    return end_effector_pose
+
+
+def get_rotational_error(target_quat: np.ndarray, source_quat: np.ndarray):
```

Comment:
Maybe call this `quat_angle` or something since it's getting the angle between two quaternions??